tests/core/startup/ErrorControlChainTest.php

```diff
@@ -1,90 +1,241 @@
 <?php
 
+/**
+ * An extension of ErrorControlChain that runs the chain in a subprocess.
+ *
+ * We need this because ErrorControlChain only suppresses uncaught fatal errors, and
+ * that would kill PHPUnit execution
+ */
+class ErrorControlChainTest_Chain extends ErrorControlChain {
+
+	function executeInSubprocess() {
+		// Get the path to the ErrorControlChain class
+		$classpath = SS_ClassLoader::instance()->getItemPath('ErrorControlChain');
+		$suppression = $this->suppression ? 'true' : 'false';
+
+		// Start building a PHP file that will execute the chain
+		$src = '<'."?php
+require_once '$classpath';
+
+\$chain = new ErrorControlChain();
+
+\$chain->setSuppression($suppression);
+
+\$chain
+";
+
+		// For each step, use reflection to pull out the call, stick in the the PHP source we're building
+		foreach ($this->steps as $step) {
+			$func = new ReflectionFunction($step['callback']);
+			$source = file($func->getFileName());
+
+			$start_line = $func->getStartLine() - 1;
+			$end_line = $func->getEndLine();
+			$length = $end_line - $start_line;
+
+			$src .= implode("", array_slice($source, $start_line, $length)) . "\n";
+		}
+
+		// Finally add a line to execute the chain
+		$src .= "->execute();";
+
+		// Now stick it in a temporary file & run it
+		$codepath = TEMP_FOLDER.'/ErrorControlChainTest_'.sha1($src).'.php';
+
+		$null = is_writeable('/dev/null') ? '/dev/null' : 'NUL';
+
+		file_put_contents($codepath, $src);
+		exec("php $codepath 2> $null", $stdout, $errcode);
+		unlink($codepath);
+
+		return array(implode("\n", $stdout), $errcode);
+	}
+}
+
 class ErrorControlChainTest extends SapphireTest {
 
+	function setUp() {
+		// Check we can run PHP at all
+		$null = is_writeable('/dev/null') ? '/dev/null' : 'NUL';
```

Comment:
On 16-03-31 04:17 PM, Damian Mooyman
      wrote:
    
    
      In tests/core/startup/ErrorControlChainTest.php:
      >  class ErrorControlChainTest extends SapphireTest {
>  
> +	function setUp() {
> +		// Check we can run PHP at all
> +		$null = is_writeable('/dev/null') ? '/dev/null' : 'NUL';

      NUL?
      —
        You are receiving this because you are subscribed to this
        thread.
        Reply to this email directly or view
          it on GitHub
      
        
          
          
        
        
      
    
    It's how you say NULL in
      French :P